File: webui/src/lib/utils/dataQuality.ts
/**
 * Data-quality checks, mirrored from `ofd/validation/data_quality.py`.
 *
 * These are the small, repetitive defects reviewers kept fixing by hand on webui
 * submissions after PR #405 — lowercase colour names among Title Case siblings (#451,
 * #452), placeholder empty strings (#453), spool rows that duplicate an earlier row
 * (#453), names with stray whitespace (#460). The Python module is the authority and
 * runs in CI; this file is what turns each rule into an inline "Fix" hint at entry
 * time, so a contributor never has to be told in review. Keep the two in lockstep.
 *
 * Every check is pure and returns `null` when there is nothing to say, so a form can
 * render it as `{#if issue}`. None of them mutate the value — the caller applies the
 * suggested fix only when the user presses the button.
 *
 * Name-derived trait detection is not repeated here: `traitSuggestions.ts` reads the
 * shared `schemas/trait_rules.json` table (via `/api/trait-rules`) and is wired into `VariantForm`.
 */

/** Words that legitimately stay lowercase inside a title-cased display name. */
const MINOR_WORDS = new Set([
	'and',
	'of',
	'the',
	'with',
	'de',
	'w/',
	'in',
	'on',
	'for',
	'a',
	'an'
]);

/** True when every significant word starts with an uppercase letter. */
export function isTitleCased(name: string): boolean {
	const words = name.split(/\s+/).filter((w) => w && /[a-zA-Z]/.test(w[0]));
	if (words.length === 0) return false;
	return words.every((w) => MINOR_WORDS.has(w.toLowerCase()) || w[0] === w[0].toUpperCase());
}

function isAllLower(name: string): boolean {
	return !!name && name === name.toLowerCase() && /[a-zA-Z]/.test(name);
}

/** Title-case a display name, leaving minor words and existing capitals alone. */
export function toTitleCase(name: string): string {
	return name
		.split(/(\s+)/)
		.map((token, index) => {
			if (!token.trim()) return token;
			// The first word is capitalised even when it is a minor word.
			if (index > 0 && MINOR_WORDS.has(token.toLowerCase())) return token;
			return token[0].toUpperCase() + token.slice(1);
		})
		.join('');
}

/**
 * A lowercase name among Title Case siblings.
 *
 * Casing is house style rather than a hard rule, so this only fires when the entity's
 * own siblings establish the convention — exactly how it was caught by hand on #451
 * ("translucent blue" beside "Purple", "Transparent") and #452 ("true red" beside
 * "Red", "Mint Green"). Returns the suggested replacement and an example sibling.
 */
export function checkNameCasing(
	name: string,
	siblingNames: string[]
): { suggestion: string; example: string } | null {
	if (!isAllLower(name)) return null;
	const example = siblingNames.find((s) => s && s !== name && isTitleCased(s));
	if (!example) return null;
	const suggestion = toTitleCase(name);
	if (suggestion === name) return null;
	return { suggestion, example };
}

/**
 * Whether a character is a lowercase *cased* letter — the JS equivalent of Python's
 * `str.islower()`. A caseless character (a digit, CJK, punctuation) is neither.
 */
function isLowerCased(ch: string): boolean {
	return ch === ch.toLowerCase() && ch !== ch.toUpperCase();
}

/** As {@link isLowerCased}, for uppercase (Python's `str.isupper()`). */
function isUpperCased(ch: string): boolean {
	return ch === ch.toUpperCase() && ch !== ch.toLowerCase();
}

/**
 * A display name that starts with a lowercase letter.
 *
 * Narrower and more certain than {@link checkNameCasing}: that one needs Title Case
 * siblings to establish a convention, so it stays silent on the first colour of a new
 * filament — which is exactly when a name typed straight off a product page ("yellow",
 * "glass fiber black", "kexcelled") lands in the tree. 126 names in the current data
 * start lowercase this way.
 *
 * A leading capital is not house style, it is how the field is read everywhere it is
 * displayed, so this fires on its own. Only the first letter is changed: Title Casing
 * the rest would mangle manufacturer names like "PLA+ eSilk".
 *
 * Deliberately silent on intercapped names — `eSUN 3D`, `iSANMATE`, `rPLA pro`,
 * `rPETG`, `ePAHT-CF`. A lowercase letter immediately followed by an uppercase one is
 * a brand's own styling, and there are 18 of them in the tree that must be left alone.
 */
export function checkNameLeadingCase(name: string): { suggestion: string } | null {
	if (!name) return null;
	// Find the first letter; a name may legitimately open with a digit or symbol
	// ("3D Gold", "+PLA"), and those say nothing about casing. `\p{L}` rather than
	// [a-zA-Z]: "Über ABS" is already capitalised, and treating Ü as a non-letter
	// would find the "b" and "fix" a real filament into "ÜBer ABS". `search` returns
	// a UTF-16 index, which is what the slices below need.
	const index = name.search(/\p{L}/u);
	if (index === -1) return null;
	// Take the whole code point, so an astral letter isn't split at its surrogate.
	const first = String.fromCodePoint(name.codePointAt(index)!);
	if (!isLowerCased(first)) return null;
	// eSUN / rPLA / iSANMATE — the manufacturer's own styling.
	const after = index + first.length;
	const next = after < name.length ? String.fromCodePoint(name.codePointAt(after)!) : '';
	if (next && isUpperCased(next)) return null;
	return { suggestion: name.slice(0, index) + first.toUpperCase() + name.slice(after) };
}

/**
 * Leading/trailing or repeated whitespace in a display name.
 *
 * #460 created a filament literally named `"Silk "`. The trailing space is invisible
 * in review, survives into every downstream consumer, and makes the entity look
 * distinct from the `"Silk"` the contributor meant.
 */
export function checkNameWhitespace(name: string): { suggestion: string; reason: string } | null {
	if (!name) return null;
	const collapsed = name.trim().replace(/\s+/g, ' ');
	if (collapsed === name) return null;
	return {
		suggestion: collapsed,
		reason:
			name !== name.trim()
				? 'has leading or trailing whitespace'
				: 'contains repeated whitespace'
	};
}

/**
 * Empty strings inside an array value.
 *
 * `certifications: [""]` (#453) reads downstream as "this filament has a
 * certification" whose name is blank; an empty array says the true thing. Returns the
 * indices to drop.
 */
export function checkPlaceholderEntries(values: unknown): number[] {
	if (!Array.isArray(values)) return [];
	return values.reduce<number[]>((indices, value, index) => {
		if (typeof value === 'string' && value.trim() === '') indices.push(index);
		return indices;
	}, []);
}

/** The `(filament_weight, diameter)` spool identity, mirroring `size_dedupe_key`. */
function sizeKey(size: Record<string, unknown>): string {
	return `${size.filament_weight ?? ''}|${size.diameter ?? ''}`;
}

/**
 * Fields that make a spool row worth keeping, ignoring canonical identity.
 *
 * Blank values say nothing an absent field doesn't, so they must not be what tells two
 * rows apart. Kept in step with `_redundant_size_fields` in `ofd/validation/data_quality.py`.
 */
function meaningfulFields(size: Record<string, unknown>): Array<[string, unknown]> {
	return Object.entries(size).filter(
		([key, value]) =>
			key !== 'uuid' &&
			key !== 'moved_from' &&
			value !== undefined &&
			value !== null &&
			!(typeof value === 'string' && value.trim() === '')
	);
}

/** True when `size` says nothing `earlier` doesn't already say. */
function isSubsumed(size: Record<string, unknown>, earlier: Record<string, unknown>): boolean {
	if (sizeKey(size) !== sizeKey(earlier)) return false;
	const earlierFields = new Map(meaningfulFields(earlier));
	return meaningfulFields(size).every(
		([key, value]) =>
			earlierFields.has(key) &&
			JSON.stringify(earlierFields.get(key)) === JSON.stringify(value)
	);
}

/**
 * Spool rows that add nothing over an earlier row.
 *
 * Spool identity is `(filament_weight, diameter)` — the pairing `merge_sizes` and
 * `record_moved_from` use — so two rows sharing it are the same offering unless one
 * carries a value the other lacks (a distinct GTIN, article number, purchase link,
 * spool geometry). #453 shipped two 1 kg / 1.75 mm rows with nothing between them.
 *
 * Returns, for each redundant row, its index and the index it duplicates.
 */
export function findRedundantSizes(
	sizes: Array<Record<string, unknown>>
): Array<{ index: number; duplicateOf: number }> {
	const found: Array<{ index: number; duplicateOf: number }> = [];
	for (let i = 0; i < sizes.length; i++) {
		const size = sizes[i];
		if (!size || typeof size !== 'object') continue;
		for (let j = 0; j < i; j++) {
			const earlier = sizes[j];
			if (earlier && typeof earlier === 'object' && isSubsumed(size, earlier)) {
				found.push({ index: i, duplicateOf: j });
				break;
			}
		}
	}
	return found;
}

/**
 * Traits that every sibling colour of a filament carries but this variant does not.
 *
 * Traits are a property of the product line far more often than of the individual
 * colour: if the other colours of a filament are all silk, or all carbon-filled, or
 * all industrially compostable, this one almost certainly is too. Measured against
 * the current tree by leave-one-out over the 1,167 filaments with three or more
 * colours, a trait held by every sibling is held by the held-out variant
 * 7,412/7,461 = 99.3% of the time — the strongest signal available at entry time,
 * and stronger than anything derivable from the name.
 *
 * That is why reviewers keep writing this comment by hand ("all existing
 * `kingroon/PLA/silk_pla/*` variants include `traits.silk: true`, so omitting it here
 * makes the variant inconsistent"). This turns it into a suggestion at entry time.
 *
 * Unanimity is required deliberately: a trait on some-but-not-all siblings is what a
 * genuinely per-colour trait looks like (glitter on two of twelve colours), and
 * suggesting those would train contributors to dismiss the panel.
 *
 * @param siblingTraitSets one entry per sibling variant, listing its true traits
 * @param own              traits already selected on the variant being edited
 * @param minSiblings      below this many siblings there is no consensus to read
 * @returns trait keys every sibling has and this variant lacks, in first-seen order
 */
export function suggestTraitsFromSiblings(
	siblingTraitSets: string[][],
	own: Set<string> | ReadonlySet<string>,
	minSiblings = 2
): string[] {
	if (siblingTraitSets.length < minSiblings) return [];

	const [first, ...rest] = siblingTraitSets;
	const suggestions: string[] = [];
	for (const key of first) {
		if (own.has(key) || suggestions.includes(key)) continue;
		if (rest.every((set) => set.includes(key))) suggestions.push(key);
	}
	return suggestions;
}

/** The shape `findSharedPurchaseLinks` needs off a variant — nothing more. */
export interface VariantWithLinks {
	id?: string;
	slug?: string;
	name?: string;
	sizes?: Array<{ purchase_links?: Array<{ url?: string }> | null } | null> | null;
}

/**
 * Purchase-link URLs copied across several colours of the same filament.
 *
 * A link shared by three or more colours is almost always the filament's generic
 * product page pasted onto every variant, rather than the colour-specific page a
 * buyer needs — the same thing the Rust validator's DuplicateLink rule reports.
 *
 * Unlike the other checks here there is no safe auto-fix: only the contributor can
 * know whether a colour-specific page exists, so callers surface this and let the
 * user navigate. That is also why this returns the offending URLs and who shares
 * them rather than a bare count — the count alone cannot point anyone at the problem.
 *
 * @param threshold how many colours must share a URL before it is worth reporting
 * @returns one entry per shared URL, each listing the variant ids and names sharing it
 */
export function findSharedPurchaseLinks(
	variants: VariantWithLinks[],
	threshold = 3
): Array<{ url: string; variantIds: string[]; variantNames: string[] }> {
	const byUrl = new Map<string, Map<string, string>>();

	for (const variant of variants) {
		const id = variant?.slug ?? variant?.id;
		if (!id) continue;
		for (const size of variant.sizes ?? []) {
			for (const link of size?.purchase_links ?? []) {
				const url = link?.url;
				if (!url) continue;
				if (!byUrl.has(url)) byUrl.set(url, new Map());
				// A colour counts once however many spool sizes repeat the link.
				byUrl.get(url)!.set(id, variant.name ?? id);
			}
		}
	}

	const shared: Array<{ url: string; variantIds: string[]; variantNames: string[] }> = [];
	for (const [url, owners] of byUrl) {
		if (owners.size < threshold) continue;
		shared.push({
			url,
			variantIds: [...owners.keys()],
			variantNames: [...owners.values()]
		});
	}
	return shared;
}

/**
 * Order-insensitive identity of a display name, mirroring `_word_multiset` in
 * `ofd/validation/data_quality.py` — so `PLA CF` and `CF PLA` land on one key, which is
 * what `ofd/scripts/deduplicate_data.py` would offer to merge.
 */
function wordMultiset(slug: string): string {
	return slug.split('_').filter(Boolean).sort().join('_');
}

/** The shape {@link findDuplicateName} needs off an existing entity — nothing more. */
export interface NamedEntry {
	name: string;
	/** Folder id, when known. Falls back to the slugified name. */
	slug?: string;
}

/**
 * A name already taken by another filament of the same brand.
 *
 * #281: filament names should be unique *within a brand*, not merely within a material.
 * #280 shipped a PLA-CF filament that had snuck into the PETG material beside the real
 * one — two folders, one product, and nothing in the tree said so. The folder id is the
 * unit of comparison rather than the display name, because that is what actually
 * collides on disk: `PLA+` and `PLA` both slugify to `pla`.
 *
 * `word-order` matches are the near-duplicates the Python rule warns about
 * (`cf_pla` vs `pla_cf`); an `exact` match in the *same* material cannot be created at
 * all, so the caller words that case more strongly.
 *
 * Normalisation mirrors `generateSlug` in `entityService`; kept inline so this module
 * stays dependency-free and pure.
 */
export function findDuplicateName<T extends NamedEntry>(
	name: string,
	existing: T[]
): { match: T; kind: 'exact' | 'word-order' } | null {
	const slugify = (value: string) =>
		value
			.toLowerCase()
			.replace(/[^a-z0-9]+/g, '_')
			.replace(/^_|_$/g, '');

	const slug = slugify(name);
	if (!slug) return null;
	const words = wordMultiset(slug);

	let wordOrder: T | null = null;
	for (const entry of existing) {
		const otherSlug = slugify(entry.slug || entry.name || '');
		if (!otherSlug) continue;
		if (otherSlug === slug) return { match: entry, kind: 'exact' };
		if (!wordOrder && wordMultiset(otherSlug) === words) wordOrder = entry;
	}
	return wordOrder ? { match: wordOrder, kind: 'word-order' } : null;
}

File: ofd/validation/data_quality.py
"""
Data-quality validation.

Rules for the small, repetitive defects that reviewers kept fixing by hand on webui
submissions after PR #405: placeholder values, duplicated spool rows, casing that
diverges from a filament's own siblings, fiber traits missing from variants whose
names announce the fiber, display names that start lowercase, filaments left with no
variants, and near-duplicate sibling names that differ only by word order.

This is the authoritative, server-side enforcement; the webui mirrors the rules in
``webui/src/lib/utils/dataQuality.ts`` and surfaces each as an inline "Fix" hint at
entry time. Keep the two in lockstep.

Every rule reuses the detector that already exists for its concept rather than
restating it:

* fiber codes            -> ``ofd/scripts/apply_fiber_traits.py`` (``detect_codes``)
* spool identity         -> ``ofd/merge.py`` (``size_dedupe_key``)
* word-swap normalization -> ``ofd/scripts/deduplicate_data.py``

Levels are chosen so a rule only blocks a merge when the data is genuinely wrong:
``placeholder_value``, ``duplicate_size_entry``, ``name_whitespace`` and
``purchase_link_storefront_root`` are ERRORs; the judgement calls
(``name_casing``, ``name_leading_case``, ``fiber_trait_missing``,
``orphan_filament``, ``sibling_near_duplicate``) are WARNINGs.

A WARNING here is advice, not a failure: ``ofd validate`` still exits 0 and reports
"All validations passed (N warnings)". Reviewers should not describe a warning as
something that breaks the build or makes data unreachable.
"""

import json
import re
from collections import Counter, defaultdict
from pathlib import Path

from ofd_validator import ValidationError, ValidationLevel

from ofd.merge import size_dedupe_key
from ofd.scripts.apply_fiber_traits import CODE_TRAITS, detect_codes

CATEGORY = "data_quality"

# Fields whose value is a human-facing display name.
_NAME_FIELDS = {
    "brand.json": "name",
    "material.json": "material",
    "filament.json": "name",
    "variant.json": "name",
    "store.json": "name",
}

# Two or more consecutive whitespace characters anywhere in a name.
_DOUBLED_SPACE = re.compile(r"\s\s")

# Words that legitimately stay lowercase inside a title-cased display name.
_MINOR_WORDS = {"and", "of", "the", "with", "de", "w/", "in", "on", "for", "a", "an"}


def _load_json(path: Path):
    try:
        with open(path, encoding="utf-8") as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError):
        return None


def _rel(path: Path, base: Path) -> str:
    """Path relative to the data dir's parent (project root) for readable output."""
    try:
        return str(path.relative_to(base))
    except ValueError:
        return str(path)


def _err(level, message: str, path: str) -> ValidationError:
    return ValidationError(level, CATEGORY, message, path)


# --- Rule: placeholder values -------------------------------------------------


def _placeholder_paths(value, prefix: str = "") -> list[str]:
    """Dotted paths of blank string values, recursing into dicts and lists.

    A blank string is never meaningful data. ``certifications: [""]`` (#453) reads
    downstream as "this filament has a certification" whose name is blank; an empty
    array (or an absent field) says the true thing. Whitespace-only counts as blank —
    ``" "`` is no more a certification name than ``""`` is, and the webui's
    ``checkPlaceholderEntries`` flags it as one too.
    """
    found: list[str] = []
    if isinstance(value, dict):
        for key, sub in value.items():
            found.extend(_placeholder_paths(sub, f"{prefix}.{key}" if prefix else key))
    elif isinstance(value, list):
        for index, sub in enumerate(value):
            found.extend(_placeholder_paths(sub, f"{prefix}[{index}]"))
    elif isinstance(value, str) and value.strip() == "" and prefix:
        found.append(prefix)
    return found


def _check_placeholders(data, file_path: Path, base: Path) -> list[ValidationError]:
    return [
        _err(
            ValidationLevel.Error,
            f"'{dotted}' is blank. Remove the field, or drop the empty entry "
            f"from the array — a blank value is not the same as no value.",
            _rel(file_path, base),
        )
        for dotted in _placeholder_paths(data)
    ]


# --- Rule: name whitespace ----------------------------------------------------


def _check_name_whitespace(data, file_path: Path, base: Path) -> list[ValidationError]:
    """Leading/trailing or doubled whitespace in a display name.

    #460 created a filament literally named ``"Silk "``. The trailing space is
    invisible in review, survives into every downstream consumer, and makes the entity
    look distinct from the ``"Silk"`` the contributor meant.
    """
    field = _NAME_FIELDS.get(file_path.name)
    if not field or not isinstance(data, dict):
        return []
    name = data.get(field)
    if not isinstance(name, str) or not name:
        return []

    if name != name.strip():
        return [
            _err(
                ValidationLevel.Error,
                f"Name {name!r} has leading or trailing whitespace.",
                _rel(file_path, base),
            )
        ]
    if _DOUBLED_SPACE.search(name):
        return [
            _err(
                ValidationLevel.Error,
                f"Name {name!r} contains repeated whitespace.",
                _rel(file_path, base),
            )
        ]
    return []


# --- Rule: duplicate size entries ---------------------------------------------


def _redundant_size_fields(size: dict) -> dict:
    """The fields that make a spool row worth keeping, ignoring canonical identity.

    A field set to ``None`` or a blank string says nothing the absent field doesn't, so
    it must not be what distinguishes two rows — otherwise ``{"gtin": null}`` reads as a
    distinct GTIN. The webui's ``meaningfulFields`` drops the same values.
    """
    return {
        k: v
        for k, v in size.items()
        if k not in ("uuid", "moved_from")
        and v is not None
        and not (isinstance(v, str) and v.strip() == "")
    }


def _is_subsumed(size: dict, earlier: dict) -> bool:
    """True when `size` says nothing `earlier` doesn't already say.

    Spool identity in this repo is ``(filament_weight, diameter)`` — see
    ``size_dedupe_key`` in ``ofd/merge.py``, which is what ``merge_sizes`` and
    ``record_moved_from`` pair on. So two rows sharing that key are the same spool, and
    the later one is redundant unless it carries a value the earlier one lacks or
    contradicts (a distinct GTIN, article number, purchase link, spool geometry...).

    #453 shipped two 1 kg / 1.75 mm rows with nothing to tell them apart, which renders
    as the same size listed twice.
    """
    if size_dedupe_key(size) != size_dedupe_key(earlier):
        return False
    earlier_fields = _redundant_size_fields(earlier)
    return all(
        key in earlier_fields and earlier_fields[key] == value
        for key, value in _redundant_size_fields(size).items()
    )


def _check_duplicate_sizes(sizes, file_path: Path, base: Path) -> list[ValidationError]:
    if not isinstance(sizes, list):
        return []

    errors: list[ValidationError] = []
    for index, size in enumerate(sizes):
        if not isinstance(size, dict):
            continue
        for earlier_index in range(index):
            earlier = sizes[earlier_index]
            if isinstance(earlier, dict) and _is_subsumed(size, earlier):
                weight = size.get("filament_weight")
                diameter = size.get("diameter")
                errors.append(
                    _err(
                        ValidationLevel.Error,
                        f"Size #{index + 1} ({weight}g, {diameter}mm) adds nothing over size "
                        f"#{earlier_index + 1}. Merge them, or give it the SKU, GTIN or "
                        f"purchase link that tells them apart.",
                        _rel(file_path, base),
                    )
                )
                break
    return errors


# --- Rule: name casing --------------------------------------------------------


def _is_title_cased(name: str) -> bool:
    """True when every significant word starts with an uppercase letter."""
    words = [w for w in name.split() if w and w[0].isalpha()]
    if not words:
        return False
    return all(w[0].isupper() for w in words if w.lower() not in _MINOR_WORDS)


def _is_all_lower(name: str) -> bool:
    return bool(name) and name == name.lower() and any(c.isalpha() for c in name)


def _check_sibling_casing(names: list[tuple[str, Path]], base: Path) -> list[ValidationError]:
    """Flag an all-lowercase name among Title Case siblings.

    Casing is a house style rather than a hard rule, so this only fires when the
    entity's own siblings establish the convention — which is exactly how it was
    caught by hand on #451 ("translucent blue" beside "Purple", "Transparent") and
    #452 ("true red" beside "Red", "Mint Green").
    """
    titled = [n for n, _ in names if _is_title_cased(n)]
    if not titled:
        return []

    return [
        _err(
            ValidationLevel.Warning,
            f"Name {name!r} is lowercase while its siblings use Title Case (e.g. {titled[0]!r}).",
            _rel(path, base),
        )
        for name, path in names
        if _is_all_lower(name)
    ]


# --- Rule: name starts lowercase ----------------------------------------------


def _check_name_leading_case(data, file_path: Path, base: Path) -> list[ValidationError]:
    """A display name that starts with a lowercase letter.

    Narrower and more certain than ``_check_sibling_casing``: that rule needs Title
    Case siblings to establish a convention, so it stays silent on the first colour
    of a new filament — which is exactly when a name pasted off a product page
    ("yellow", "glass fiber black") lands in the tree. 126 names in the current data
    start lowercase this way.

    Only the first letter is at issue; Title Casing the rest would rewrite
    manufacturer styling like "PLA+ eSilk". Intercapped names are left alone —
    "eSUN 3D", "iSANMATE", "rPLA pro", "rPETG", "ePAHT-CF" are the brand's own
    styling, and there are 18 such names in the tree.

    Mirrors ``checkNameLeadingCase`` in ``webui/src/lib/utils/dataQuality.ts``.
    """
    field = _NAME_FIELDS.get(file_path.name)
    if not field or not isinstance(data, dict):
        return []
    name = data.get(field)
    if not isinstance(name, str) or not name:
        return []

    # A name may legitimately open with a digit or symbol ("3D Gold", "+PLA"); those
    # say nothing about casing, so look at the first letter wherever it is.
    index = next((i for i, ch in enumerate(name) if ch.isalpha()), None)
    if index is None or not name[index].islower():
        return []
    # eSUN / rPLA / iSANMATE.
    if index + 1 < len(name) and name[index + 1].isupper():
        return []

    suggestion = name[:index] + name[index].upper() + name[index + 1 :]
    return [
        _err(
            ValidationLevel.Warning,
            f"Name {name!r} starts with a lowercase letter; display names are shown "
            f"capitalised. Use {suggestion!r}.",
            _rel(file_path, base),
        )
    ]


# --- Rule: fiber traits missing -----------------------------------------------


def _check_fiber_traits(
    data: dict, text: str, file_path: Path, base: Path
) -> list[ValidationError]:
    """A variant whose name announces CF/GF/HF but carries none of the traits.

    ``apply_fiber_traits.py`` has detected this since #405, but only to *suggest*.
    #450 still merged four variants named "carbon fiber <colour>" with no
    ``contains_carbon_fiber`` or ``abrasive``, so they were invisible to every
    downstream abrasive-material filter.
    """
    codes = detect_codes(text)
    if not codes:
        return []

    traits = data.get("traits")
    traits = traits if isinstance(traits, dict) else {}

    missing: list[str] = []
    for code in sorted(codes):
        for trait in CODE_TRAITS.get(code, ()):
            if traits.get(trait) is not True and trait not in missing:
                missing.append(trait)

    if not missing:
        return []

    return [
        _err(
            ValidationLevel.Warning,
            f"Name suggests {', '.join(sorted(codes))} but these traits are not set: "
            f"{', '.join(missing)}.",
            _rel(file_path, base),
        )
    ]


# --- Rule: orphan filament ----------------------------------------------------


def _check_orphan_filament(filament_dir: Path, base: Path) -> list[ValidationError]:
    """A filament directory with no variants under it.

    This flags an *incomplete* filament, not a broken one. Childless entities are
    valid, addressable and rendered: the API emits an ``index.json`` for them and
    the webui's filament page loads the filament independently of its variants, so
    a filament with no colours shows an empty Variants panel rather than a 404.
    There are ~140 of them in the tree at any time, most of them brand skeletons
    waiting on a follow-up PR.

    It is worth reporting because it is also what a botched edit looks like: #461
    left ``data/3dhojor/PLA/silk_blue_green/`` behind this way, by deleting and
    re-creating a filament that a still-open PR was also editing.

    Hence WARNING, and hence the wording below leads with adding colours rather
    than with deletion — do not restate this rule as "unreachable in the UI".
    """
    has_variant = any(
        child.is_dir() and (child / "variant.json").exists() for child in filament_dir.iterdir()
    )
    if has_variant:
        return []
    return [
        _err(
            ValidationLevel.Warning,
            "Filament has no colours yet. Add at least one variant when you have the "
            "data; if this folder is a leftover from an edit, remove it.",
            _rel(filament_dir, base),
        )
    ]


# --- Rule: sibling near-duplicates --------------------------------------------


def _word_multiset(name: str) -> tuple:
    """Order-insensitive identity of an underscore-separated id.

    Mirrors the word-swap grouping in ``ofd/scripts/deduplicate_data.py``, so what
    that script would offer to merge is what this reports.
    """
    return tuple(sorted(Counter(name.split("_")).items()))


def _check_sibling_duplicates(ids: list[tuple[str, Path]], base: Path) -> list[ValidationError]:
    groups: dict[tuple, list[tuple[str, Path]]] = defaultdict(list)
    for entity_id, path in ids:
        groups[_word_multiset(entity_id)].append((entity_id, path))

    errors: list[ValidationError] = []
    for members in groups.values():
        if len(members) < 2:
            continue
        names = [entity_id for entity_id, _ in members]
        for entity_id, path in members[1:]:
            others = [n for n in names if n != entity_id]
            errors.append(
                _err(
                    ValidationLevel.Warning,
                    f"'{entity_id}' is a word-order duplicate of {', '.join(repr(o) for o in others)}. "
                    f"These are almost always the same thing under two names.",
                    _rel(path, base),
                )
            )
    return errors


# --- Entry point --------------------------------------------------------------


def check_data_quality(data_dir) -> list[ValidationError]:
    """Scan the data tree and report every data-quality finding."""
    data_dir = Path(data_dir)
    errors: list[ValidationError] = []
    if not data_dir.exists():
        return errors

    base = data_dir.parent

    for brand_dir in sorted(p for p in data_dir.iterdir() if p.is_dir()):
        brand_file = brand_dir / "brand.json"
        brand_data = _load_json(brand_file)
        if isinstance(brand_data, dict):
            errors.extend(_check_placeholders(brand_data, brand_file, base))
            errors.extend(_check_name_whitespace(brand_data, brand_file, base))
            errors.extend(_check_name_leading_case(brand_data, brand_file, base))

        for material_dir in sorted(p for p in brand_dir.iterdir() if p.is_dir()):
            material_file = material_dir / "material.json"
            material_data = _load_json(material_file)
            if isinstance(material_data, dict):
                errors.extend(_check_placeholders(material_data, material_file, base))
                errors.extend(_check_name_whitespace(material_data, material_file, base))
                errors.extend(_check_name_leading_case(material_data, material_file, base))

            filament_names: list[tuple[str, Path]] = []
            filament_ids: list[tuple[str, Path]] = []

            for filament_dir in sorted(p for p in material_dir.iterdir() if p.is_dir()):
                filament_file = filament_dir / "filament.json"
                filament_data = _load_json(filament_file)
                if not isinstance(filament_data, dict):
                    continue

                errors.extend(_check_placeholders(filament_data, filament_file, base))
                errors.extend(_check_name_whitespace(filament_data, filament_file, base))
                errors.extend(_check_name_leading_case(filament_data, filament_file, base))
                errors.extend(_check_orphan_filament(filament_dir, base))

                name = filament_data.get("name")
                if isinstance(name, str) and name:
                    filament_names.append((name, filament_file))
                filament_ids.append((filament_dir.name, filament_dir))

                variant_names: list[tuple[str, Path]] = []
                variant_ids: list[tuple[str, Path]] = []

                for variant_dir in sorted(p for p in filament_dir.iterdir() if p.is_dir()):
                    variant_file = variant_dir / "variant.json"
                    variant_data = _load_json(variant_file)
                    if not isinstance(variant_data, dict):
                        continue

                    errors.extend(_check_placeholders(variant_data, variant_file, base))
                    errors.extend(_check_name_whitespace(variant_data, variant_file, base))
                    errors.extend(_check_name_leading_case(variant_data, variant_file, base))

                    # Fiber codes can come from any level of the path, matching
                    # apply_fiber_traits.py's scan.
                    fiber_text = " / ".join(
                        str(part).lower()
                        for part in (
                            material_dir.name,
                            filament_dir.name,
                            filament_data.get("name") or "",
                            variant_dir.name,
                            variant_data.get("name") or "",
                        )
                        if part
                    )
                    errors.extend(_check_fiber_traits(variant_data, fiber_text, variant_file, base))

                    variant_name = variant_data.get("name")
                    if isinstance(variant_name, str) and variant_name:
                        variant_names.append((variant_name, variant_file))
                    variant_ids.append((variant_dir.name, variant_dir))

                    sizes_file = variant_dir / "sizes.json"
                    if sizes_file.exists():
                        sizes = _load_json(sizes_file)
                        errors.extend(_check_placeholders(sizes, sizes_file, base))
                        errors.extend(_check_duplicate_sizes(sizes, sizes_file, base))

                errors.extend(_check_sibling_casing(variant_names, base))
                errors.extend(_check_sibling_duplicates(variant_ids, base))

            errors.extend(_check_sibling_casing(filament_names, base))
            errors.extend(_check_sibling_duplicates(filament_ids, base))

    return errors
